### ISSUE TITLE
Rule S4150: False positive on field used in switch block

### ISSUE THREAD

**Issue by antoinebj** [SonarSource/sonar-csharp#1434]
### Description

Sonar warns of a violation of rule S1450 when the following conditions are met:
- A field is set in the constructor of a class
- The field is only read within the switch block of a method of that class
- The switch block switches on a parameter of an enum type that is defined in a separate file (but not necessarily separate namespace)

### Repro steps

1. Add the following code to a project:

Class1.cs:
```csharp
internal class Class1
{
    private readonly int value;

    public Class1(int value)
    {
        this.value = value;
    }

    public int GetValue(Enum1 type)
    {
        switch (type)
        {
            case Enum1.EnumValue2:
                return this.value;
            default:
                throw new NotSupportedException();
        }
    }
}
```
Enum1.cs:
```csharp
internal enum Enum1
{
    EnumValue1,
    EnumValue2,
    EnumValue3
}
```

2. Run sonar analysis on that code

### Expected behavior

There should be no warning of a violation of rule S1450.

### Actual behavior

There is a warning of a violation of rule S1450.

Note that if Enum1 is defined just above the class, in the same file and same namespace block, the warning does not happen.
If the switch case is replace with `if`, the warning also does not happen.

### Related information

* SonarC# Version : 7.1
* Visual Studio Version : 15.7

**Comment by valhristov** [comment]
Hi @antoinebj the problem appears because the enum is declared in another file... The analyzer tries to recompile modified versions of the code where each field is replaced by a local variable and then looks for a particular error (use of not initialized variable) to determine if the field is removable or not. When the code cannot be compiled for other reasons (missing enum declaration in your sample) the error we are looking for does not appear and we generate false positive.

We just replaced the implementation of S1450 and this problem should disappear. In general this rule should now generate much less false alarms.
